### PR TITLE
Add keyFn for Templates

### LINKED ISSUE #249: Form items will unmount after form rerendering.
### Issue Type

'bug report'
### Description

The issue is related to the Form

#### Steps to Reproduce

1) Open the demo and the developer console.
2) Wait 3 sec and see that the form item component will unmount after setState in Example component

#### Current Behavior
Form item component will unmount after rerendering.
#### Expected Behavior
Expected that the form item will not unmount or even will not rerender (if item props the same) after setState in Example component.
#### Demo

https://stackblitz.com/edit/devextreme-react-ft5hi8?file=index.js
### Environment Details

- devextreme version: "^18.2.2-pre-beta"
- devextreme-react version: "^18.2.3"
- react version: "^16.6.0"

### PR DESCRIPTION
**Pull request by lukyanovas**
partially resolves #249

### CODE REVIEW

File: src/core/__tests__/template.test.tsx [hunk 1]
```diff
@@ -336,6 +352,22 @@ describe("nested template", () => {
         expect(component.find(".template").html()).toBe('<div class="template">Second Template</div>');
     });
 
+    it("has templates in state with ids genetated with keyExpr", () => {
```
Comment (IlyaKhD) [inline src/core/__tests__/template.test.tsx]: Both added tests have the same name

File: src/core/template-host.ts [hunk 2]
```diff
@@ -126,10 +127,14 @@ class TemplateHost {
...
-function wrapTemplate(createContentProvider: () => (model: any) => any, stateUpdater: StateUpdater): IDxTemplate {
+function wrapTemplate(
+    createContentProvider: () => (model: any) => any,
+    stateUpdater: StateUpdater,
+    keyExpr?: (data) => string
+): IDxTemplate {
     return {
         render: (data: IDxTemplateData) => {
-            const templateId = "__template_" + generateID();
+            const templateId = keyExpr && keyExpr(data.model) || "__template_" + generateID();
```
Comment (IlyaKhD) [inline src/core/template-host.ts]:
```suggestion
            const templateId = keyExpr ? keyExpr(data.model) : "__template_" + generateID();
```

File: tools/src/component-generator.ts [hunk 3]
```diff
@@ -204,7 +208,8 @@ function createTemplateDto(templates: string[]) {
     ? templates.map((actualOptionName) => ({
         actualOptionName,
         render: formatTemplatePropName(actualOptionName, "Render"),
-        component: formatTemplatePropName(actualOptionName, "Component")
+        component: formatTemplatePropName(actualOptionName, "Component"),
+        keyExpr: formatTemplatePropName(actualOptionName, "ReactKeyExpr")
```
Comment (IlyaKhD) [inline tools/src/component-generator.ts]: Could we use smth shorter and simplier then `ReactKeyExpr` suffix?
`Expr` part is unnecessary and may be confusing as Expressions are used widely in the DevExtreme data binding.

What about `ReactKey`.. well, it's shorter, but personally I don't like using word React in this case. It's all React)

May be `ElementKey` or `RenderKey` or smthLike ?

Comment (lukyanovas) [reply tools/src/component-generator.ts]: We add smth like key getter, not a key

File: src/core/__tests__/template.test.tsx (hunk 1 above)
Comment (lukyanovas) [reply src/core/__tests__/template.test.tsx]: > It's tested in different suits

**Pull request merged**